fhirpipe_clean/extract/sql.py

```diff
@@ -0,0 +1,292 @@
+    columns = []
+    joins = set()
+    if isinstance(tree, dict):
+        if "inputColumns" in tree.keys() and tree["inputColumns"]:
+            for col in tree["inputColumns"]:
+                # Check if table and column target are defined
+                if col["table"] is not None and col["column"] is not None:
+                    column_name = build_col_name(col["table"], col["column"], col["owner"])
+                    columns.append(column_name)
+
+                # If there is a join
+                if "joins" in col.keys() and len(col["joins"]) > 0:
+                    for join in col["joins"]:
+                        # Add the join
+                        source_col = build_col_name(join["sourceTable"], join["sourceColumn"], join["sourceOwner"])
+                        target_col = build_col_name(join["targetTable"], join["targetColumn"], join["targetOwner"])
+                        joins.add((source_col, target_col))
+                # Else it's a static value and there is nothing to do
+
+        else:  # Recurse in child
+            cols_child, joins_child = find_cols_and_joins(tree["attributes"])
+            columns += cols_child
+            joins = joins.union(joins_child)
+
+    elif isinstance(tree, list) and len(tree) > 0:
+        for t in tree:
+            cols_children, joins_children = find_cols_and_joins(t)
+            columns += cols_children
+            joins = joins.union(joins_children)
+
+    return columns, joins
+
+
+def build_graph(joins):
+    """
+    Transform a join info into SQL fragments and parse the graph of join dependency
+    Input:
+        [(<type_of_join>, "<owner>.<table>.<col>=<owner>.<join_table>.<join_col>"), ... ]
+    Return:
+        [(
+            "<join_table>",
+            "<table>.<col> = <join_table>.<join_col>"
+        ), ... ],
+        graph of dependency of type DependencyGraph
+    """
+    print("build graph")
+    print(joins)
+    joins_elems = dict()
+    graph = DependencyGraph()
+    for join in joins:
+        join_source, join_target = join
+
+        # Get table names
+        target_table = get_table_name(join_target)
+        source_table = get_table_name(join_source)
+
+        # Add the join in the join_graph
+        source_node = graph.get(source_table)
+        target_node = graph.get(target_table)
+        source_node.connect(target_node)
+    return graph
+
+
+def build_squash_rule(node, table_col_idx):
+    """
+    Using the dependency graph of the joins on the tables (accessed through the
+    head node), regroup (using the id) the columns which should be squashed (ie
+    those accessed through a OneToMany join)
+    :param node: the node of the source table (which can be relative in recursive calls)
+    :param table_col_idx: a dict [table_name]: list of idx of cols in the SQL response
+        which come from this table
+    :return: [
+        (idx cols for source_table),
+        [
+            (idx cols for join OneToMany n1, []),
```

Comment:
Could you add `...` -> `(idx cols for join OneToMany n1, [...]),` so that the fact that it is recurrent is more clear?